### DISCUSSION TITLE
How to solve the problem of data imbalance in the training of Yolov8-SEG model?

### DISCUSSION THREAD

**Discussion by Eternal-passion** [ultralytics/ultralytics#20316] [Q&A]
I encountered a serious data imbalance during training, but I didn't find a class-weighted parameter interface, what specific aspects can I solve this problem?
![b9a71a949024c130aecea1a28e11707](https://github.com/user-attachments/assets/b644f2f5-6be8-48b3-a194-fd71a1add863)
![MaskPR_curve](https://github.com/user-attachments/assets/2d9a8b52-1554-4862-8709-50dfa42f20ec)
![BoxPR_curve](https://github.com/user-attachments/assets/581f4304-2e53-4b0f-b33d-e756e84301bf)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Eternal-passion, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://reddit.com/r/Ultralytics) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

🔔 This is an automated response. An Ultralytics engineer will review your question and assist you soon!

**Reply by glenn-jocher** [reply]
Hi @Eternal-passion, dealing with class imbalance is indeed challenging in segmentation tasks. While YOLOv8 doesn't have a direct class-weighted loss parameter, you can try several approaches:

1. Data augmentation focused on minority classes
2. Oversampling instances of underrepresented classes in your dataset
3. Using the `--rect` flag to preserve aspect ratios which can help with small objects
4. Adjusting hyperparameters like `cls` (classification loss weight) and `iou` (IOU loss weight) in your config

For implementation, you can modify the loss function in the YOLOv8 source code (look at `ultralytics/models/yolo/detect/loss.py` or the equivalent segmentation loss file) to incorporate class weights based on your distribution.

If you'd like more detailed guidance, please share your class distribution and training configuration.

**Reply by Eternal-passion** [reply]
If I want to solve this problem by modifying the loss function, how should I modify it, and do I need to load a custom model file when I train?

**Reply by glenn-jocher** [reply]
Hi @Eternal-passion, to implement class weighting in the segmentation loss function, you'll need to modify the Ultralytics source code. Here's how to approach this:

1. Create a custom loss file by copying `ultralytics/utils/loss.py` to your project directory
2. Modify the `v8SegmentationLoss` class to add class weighting in the `__call__` method
3. Specifically, adjust the BCE loss calculation line (around line 319):
   ```python
   # Original
   loss[2] = self.bce(pred_scores, target_scores.to(dtype)).sum() / target_scores_sum  # BCE
   
   # Modified with class weights (example)
   class_weights = torch.tensor([w1, w2, w3, ...], device=self.device)  # your class weights
   weighted_loss = self.bce(pred_scores, target_scores.to(dtype)) * class_weights[target_labels]
   loss[2] = weighted_loss.sum() / target_scores_sum
   ```

You can also implement the FocalLoss approach shown in the reference material which has built-in parameters (gamma and alpha) to handle class imbalance.

After modifying, you'll need to register your custom loss function and train with:
```python
from ultralytics import YOLO
from your_custom_loss import CustomSegLoss

model = YOLO('yolov8n-seg.yaml')
model.loss = CustomSegLoss(model)
model.train(data='your_data.yaml')
```

This approach requires more code changes but gives you precise control over class weighting.

**Reply by Eternal-passion** [reply]
What does the target_labels parameter mean here, and how should I give this parameter

**Reply by glenn-jocher** [reply]
Hi @Eternal-passion, I apologize for the confusion. Looking more closely at the `v8SegmentationLoss` implementation, `target_labels` isn't directly available in that context. Let me clarify how you should implement class weighting:

When examining the loss function in `ultralytics/utils/loss.py`, the BCE loss is calculated using `pred_scores` and `target_scores`:

```python
loss[2] = self.bce(pred_scores, target_scores.to(dtype)).sum() / target_scores_sum  # BCE
```

To apply class weights, you'd need to modify this to use the class indices from the ground truth. A more accurate implementation would be:

```python
# Define your class weights based on your dataset distribution
class_weights = torch.tensor([1.0, 2.0, 3.0, ...], device=self.device)  # example weights

# Extract class indices from batch 
cls_indices = batch["cls"].long().flatten()

# Get weights for each instance
weights = class_weights[cls_indices]

# Apply weights to BCE loss
weighted_loss = self.bce(pred_scores, target_scores.to(dtype)) * weights.view_as(target_scores)
loss[2] = weighted_loss.sum() / target_scores_sum
```

You'll need to adjust this based on the exact tensor shapes in your implementation. The key is to map from class indices to your predefined class weights and then apply those weights to the loss values before summing.

The class weights should be determined based on the inverse frequency of each class in your dataset to give more importance to underrepresented classes.

**Reply by Eternal-passion** [reply]
I got a mismatch between weights and target_scores size during training (my batch_size value is 24 and the number of categories is 5):
target_scores shape: torch. Size([24, 14196, 5])numel: 1703520
weights shape: torch. Size([195]), numel: 195
What is the problem and how can I fix it? Another point is that I still want to know what the target_labels argument means, and I see that there are many comments in the code that use varifocal_loss or focal_loss, and I wonder if I can use focal_loss to solve this long-tail problem, and if focal_loss so, how should my code be changed?

**Reply by glenn-jocher** [reply]
Hi @Eternal-passion, I see the dimension mismatch issue in your implementation. Let me address both this and your questions about focal loss:

The shape mismatch happens because `target_scores` has shape [24, 14196, 5] while your weights are just a flat tensor of shape [195]. To properly apply class weights, you need to match the dimensions:

```python
# Define class weights for your 5 classes
class_weights = torch.tensor([1.0, 2.0, 3.0, 4.0, 5.0], device=self.device)  # example weights

# Looking at the segmentation loss code, we need to expand weights to match target_scores
expanded_weights = class_weights.view(1, 1, 5).expand(24, 14196, 5)
# Now expanded_weights has shape [24, 14196, 5] matching target_scores

weighted_loss = self.bce(pred_scores, target_scores.to(dtype)) * expanded_weights
loss[2] = weighted_loss.sum() / target_scores_sum
```

Regarding focal loss - yes, this is actually a better approach for class imbalance! The `FocalLoss` class in `ultralytics/utils/loss.py` is specifically designed to address this problem by down-weighting well-classified examples.

To use focal loss instead of BCE, modify the segmentation loss function:

```python
# Replace the BCE loss with focal loss
# self.focal = FocalLoss(gamma=2.0, alpha=0.25)  # Define this in __init__
loss[2] = self.focal(pred_scores, target_scores) # FocalLoss already handles the sum
```

The `alpha` parameter in FocalLoss can be a tensor with per-class weights matching your class distribution. This is more elegant than manually applying weights to BCE loss and likely to give better results for your imbalanced dataset.
